tests/test_numpy_wrapper.py

```diff
@@ -0,0 +1,293 @@
+        assert not x.requires_grad
+
+        x = fn([1, 1, 2], requires_grad=False, **kwargs)
+        assert not x.requires_grad
+
+    @pytest.mark.parametrize("fn, kwargs", ARRAY_SHAPE_FNS)
+    def test_tensor_creation_from_shape(self, fn, kwargs):
+        """Test that you can create the tensor class from NumPy functions
+        instantiated via shapes with the requires_grad argument"""
+        # default value is true
+        shape = 4
+        x = fn(shape, **kwargs)
+
+        assert isinstance(x, np.tensor)
+        assert x.requires_grad
+
+        x = fn(shape, requires_grad=True, **kwargs)
+        assert x.requires_grad
+
+        x.requires_grad = False
+        assert not x.requires_grad
+
+        x = fn(shape, requires_grad=False, **kwargs)
+        assert not x.requires_grad
+
+    def test_tensor_creation_from_string(self):
+        """Test that a tensor is properly created from a string."""
+        string = "5, 4, 1, 2"
+        x = np.fromstring(string, dtype=int, sep=',')
+
+        assert isinstance(x, np.tensor)
+        assert x.requires_grad
+
+        x = np.fromstring(string, requires_grad=True, dtype=int, sep=',')
+        assert x.requires_grad
+
+        x.requires_grad = False
+        assert not x.requires_grad
+
+        x = np.fromstring(string, requires_grad=False, dtype=int, sep=',')
+        assert not x.requires_grad
+
+    def test_wrapped_docstring(self, capsys):
+        """Test that wrapped NumPy functions retains the original
+        docstring."""
+        print(np.sin.__doc__)
+        captured = capsys.readouterr()
+        assert "Trigonometric sine, element-wise." in captured.out
+
+    def test_wrapped_function_on_tensor(self):
+        """Test that wrapped functions work correctly"""
+        x = np.array([0, 1, 2], requires_grad=True)
+        res = np.sin(x)
+        expected = onp.sin(onp.array([0, 1, 2]))
+        assert np.all(res == expected)
+        assert res.requires_grad
+
+        # since the wrapping is dynamic, even ``sin`` will
+        # now accept the requires_grad keyword argument.
+        x = np.array([0, 1, 2], requires_grad=True)
+        res = np.sin(x, requires_grad=False)
+        expected = onp.sin(onp.array([0, 1, 2]))
+        assert np.all(res == expected)
+        assert not res.requires_grad
+
+    def test_wrapped_function_nontrainable_input(self):
+        """Test that a wrapped function acting on non-trainable
+        input returns non-trainable output"""
+        arr1 = np.array([0, 1], requires_grad=False)
+        arr2 = np.array([2, 3], requires_grad=False)
+        arr3 = np.array([4, 5], requires_grad=False)
+
+        res = np.vstack([arr1, arr2, arr3])
+        assert not res.requires_grad
+
+        # If one of the inputs is trainable, the output always is.
+        arr1.requires_grad = True
+        res = np.vstack([arr1, arr2, arr3])
+        assert res.requires_grad
+
+    def test_wrapped_function_nontrainable_variable_args(self):

```

Comment:
I didn't quite get what this test did that the one above wasn't doing :thinking:

Reply:
Ah, I added in this test last minute, as I wanted to check if the wrapping worked for NumPy functions of the form `np.func([arr1, arr2, ...])`, as well as `np.func(arr1, arr2)` 😆 Essentially, just wanted to make sure all bases were covered. Will make this clearer in the docstring.

Reply:
Have updated the docstrings and the test names